From 8830cdabf0d839bf64711c8c1ec8d15ebd1260a3 Mon Sep 17 00:00:00 2001
From: Rimon Xu <rimon.xu@rock-chips.com>
Date: Fri, 21 Dec 2018 14:32:07 +0800
Subject: [PATCH] driver: video: rockchip: set vepu clk in power on/off.

In order to achieve 1080P30 performance in the presence of MJPEG
decoding and H264 encoding, it is necessary to raise the clk_core
to 300M.

Change-Id: Ifc9861d02694ce55d1176d5d4ad41c40a1d4e034
Signed-off-by: Rimon Xu <rimon.xu@rock-chips.com>
---
 drivers/video/rockchip/vpu/mpp_dev_vepu.c | 25 ++++++++++++++++++-----
 1 file changed, 20 insertions(+), 5 deletions(-)

diff --git a/drivers/video/rockchip/vpu/mpp_dev_vepu.c b/drivers/video/rockchip/vpu/mpp_dev_vepu.c
index 18e470c46755..5e97eb06ffc3 100644
--- a/drivers/video/rockchip/vpu/mpp_dev_vepu.c
+++ b/drivers/video/rockchip/vpu/mpp_dev_vepu.c
@@ -40,6 +40,13 @@
 #define		VEPU_ENC_FMT_VP8E		1
 #define		VEPU_ENC_ENABLE			BIT(0)
 
+#define MHZ					(1000 * 1000)
+
+#define DEFAULT_ACLK_RATE_ON			(200 * MHZ)
+#define DEFAULT_CCLK_RATE_ON			(300 * MHZ)
+#define DEFAULT_ACLK_RATE_OFF                   (50 * MHZ)
+#define DEFAULT_CCLK_RATE_OFF                   (50 * MHZ)
+
 /*
  * file handle translate information
  */
@@ -283,24 +290,32 @@ static void rockchip_mpp_vepu_power_on(struct rockchip_mpp_dev *mpp)
 {
 	struct rockchip_vepu_dev *enc = to_vepu_dev(mpp);
 
-	if (enc->aclk)
+	if (enc->aclk) {
+		clk_set_rate(enc->aclk, DEFAULT_ACLK_RATE_ON);
 		clk_prepare_enable(enc->aclk);
+	}
 	if (enc->hclk)
 		clk_prepare_enable(enc->hclk);
-	if (enc->cclk)
+	if (enc->cclk) {
+		clk_set_rate(enc->cclk, DEFAULT_CCLK_RATE_ON);
 		clk_prepare_enable(enc->cclk);
+	}
 }
 
 static void rockchip_mpp_vepu_power_off(struct rockchip_mpp_dev *mpp)
 {
 	struct rockchip_vepu_dev *enc = to_vepu_dev(mpp);
 
+	if (enc->aclk) {
+		clk_set_rate(enc->aclk, DEFAULT_ACLK_RATE_OFF);
+		clk_disable_unprepare(enc->aclk);
+	}
 	if (enc->hclk)
 		clk_disable_unprepare(enc->hclk);
-	if (enc->aclk)
-		clk_disable_unprepare(enc->aclk);
-	if (enc->cclk)
+	if (enc->cclk) {
+		clk_set_rate(enc->cclk, DEFAULT_CCLK_RATE_OFF);
 		clk_disable_unprepare(enc->cclk);
+	}
 }
 
 static int rockchip_mpp_vepu_probe(struct rockchip_mpp_dev *mpp)
-- 
2.35.3

